addons/l10n_es_edi_verifactu/models/verifactu_document.py

```diff
@@ -0,0 +1,1186 @@
+                tax_amount = tax_detail['tax_amount']
+                if float_round(tax_percentage, precision_digits=2) or float_round(tax_amount, precision_digits=2):
+                    errors.append(_("No Sujeto VAT taxes must have 0 amount."))
+            if len(tax_detail['recargo_taxes']) > 1:
+                errors.append(_("Only a single recargo tax may used per \"main\" tax."))
+
+        verifactu_tax_types = {
+            tax_detail['verifactu_tax_type']
+            for tax_detail in tax_details['tax_details'].values()
+            if tax_detail['is_main_tax']
+        }
+        if len(verifactu_tax_types) > 1:
+            name_map = self.env['account.tax']._l10n_es_edi_verifactu_get_tax_types_name_map()
+            human_readable_types = [name_map[t] for t in verifactu_tax_types]
+            errors.append(_("We only allow a single Veri*Factu Tax Type per document: %(types)s.",
+                            types=', '.join(human_readable_types)))
+
+        for record_detail in tax_details['tax_details_per_record'].values():
+            main_tax_details = [tax_detail for key, tax_detail in record_detail['tax_details'].items() if key['is_main_tax']]
+            if len(main_tax_details) > 1 or any(len(tax_detail['group_tax_details']) > 1 for tax_detail in main_tax_details):
+                errors.append(_("We only allow a single \"main\" tax per line."))
+                # Giving the errors once should be enough
+                break
+
+        return errors
+
+    def _create_for_record(self, record_values, previous_record_identifier=None):
+        """Note: In case we succesfully create a JSON we delete all linked documents that failed the JSON creation."""
+        company = record_values['company']
+        document_vals = record_values['document_vals']
+        generation_errors = record_values['errors']
+
+        json_string = None
+        if generation_errors:
+            error_title = _("The Veri*Factu document could not be created")
+            document_vals['errors'] = self._format_errors(error_title, generation_errors)
+        else:
+            render_vals = self._render_vals(
+                record_values, previous_record_identifier=previous_record_identifier,
+            )
+            # We do not allow generating documents that would change the record identifier (i.e. values in the QR code)
+            record_identifier = render_vals['record_identifier']
+            old_record_identifier = record_values['record_identifier']
+            if old_record_identifier:
+                keys_to_check = ['IDEmisorFactura', 'NumSerieFactura', 'FechaExpedicionFactura', 'ImporteTotal']
+                changed_identifiers = {
+                    key: (old_record_identifier[key], record_identifier[key])
+                    for key in keys_to_check
+                    if old_record_identifier[key] != record_identifier[key]
+                }
+                if changed_identifiers:
```

Comment:
if we can not reset to draft, in principle they can still change the vat of the partner...  (do we need to keep everything here?)